### PR TITLE
Oleksandr Liashenko

### PR DESCRIPTION
**Pull request by oleksandr-liashenko**

### CODE REVIEW

File: HW/Oleksandr Liashenko/HW03_Task1.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+56.Namespaces are one honking great idea -- let's do more of those!"""
+# Converting text to UPPERCASE
+zen_up = zen.upper()
+# Printing how many timed does the words occur in text
+print("Times the word \"better\" occurs:", zen_up.count('BETTER'))
+print("Times the word \"never\" occurs:", zen_up.count('NEVER'))
+print("Times the word \"is\" occurs:", zen_up.count('IS'))
+# Printing text in uppercase
+print(zen_up)
+# Printing text replacing symbols
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW03_Task1.py:66]: Good job!

File: HW/Oleksandr Liashenko/HW03_Task2.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+d2 = int(num_str[2])
+d3 = int(num_str[3])
+# Product
+print("Product =", d0 * d1 * d2 * d3)
+# Given number in reverse order
+print("Number in reverse order:", num_str[3] + num_str[2] + num_str[1] + num_str[0])
+# Creating a list to sort digits
+num_list = [d0, d1, d2, d3]
+# Printing sorted digits included in the given number
+print(sorted(num_list))
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW03_Task2.py:17]: Nice

File: HW/Oleksandr Liashenko/HW03_Task3.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+b = 3
+
+print("a =", a)
+print("b =", b)
+
+a = a + b
+b = a - b
+a = a - b
+
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW03_Task3.py:13]: Correct

File: HW/Oleksandr Liashenko/HW04_Task1.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# I couldn't do it without recursion.
+
+# Solution 1. Cheating :)
+# import math
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW04_Task1.py:4]: :)

File: HW/Oleksandr Liashenko/HW04_Task1.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+# number = int(input("Enter number: "))
+# print(math.factorial(number))
+
+# Solution 2. With recursion
+number = int(input("Enter a number: "))
+result = 1
+for i in range(1, number + 1, 1):
+    result *= i
+print(result)
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW04_Task1.py:13]: Good

File: HW/Oleksandr Liashenko/HW05_Task1.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Creating a list with integers
+int_list = [3, 45, 0, 99, 66, -9, 45, 32, -101]
+# Creating an empty list for float numbers
+float_list = []
+# A loop that converts integer to float and appends to the list with float numbers
+for i in int_list:
+    float_list.append(float(i))
+# Print to check the results
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW05_Task1.py:8]: it's create new, but we need replace

File: HW/Oleksandr Liashenko/HW05_Task2.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+# Defining first two numbers of Fibonacci sequence
+x, y = 0, 1
+# Declaring a list ot write the sequence in
+Fibo = [x, y]
+# A loop which calculates next number of the sequence and while result is less than previously inputted number
+while Fibo[x]+Fibo[y] < int(number):
+    Fibo.append(Fibo[x]+Fibo[y])
+    x += 1
+    y += 1
+# Outputs result
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW05_Task2.py:12]: interesting reuse variable

File: HW/Oleksandr Liashenko/HW06_Task1.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+odd_div3 = set()
+# A set for all numbers in range
+full_range = set()
+# A loop to determine numbers and fill our sets
+for i in range(1, 11, 1):
+    full_range.add(i)
+    if i % 2 == 0:
+        even_div2.add(i)
+    if i % 3 == 0:
+        odd_div3.add(i)
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW06_Task1.py:13]: some even can be here

File: HW/Oleksandr Liashenko/HW06_Task2.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+while True:
+    login = input("Please, enter your login: ")
+    if login == "First":
+        print('Hello, First')
+        break
+    else:
+        print("Error, try again")
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW06_Task2.py:7]: good

File: HW/Oleksandr Liashenko/HW07_Task1.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+def largest(x, y):
+    """This function returns the largest number of two given numbers"""
+    if x >= y:
+        return x
+    else:
+        return y
+
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW07_Task1.py:8]: good

File: HW/Oleksandr Liashenko/HW07_Task2.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        print("Area:", s_triangle(trian_b, trian_h))
+        break
+    elif figure == 3:
+        print("Calculating area of the circle")
+        circle_rad = int(input("Enter radius: "))
+        print("Area:", s_circle(circle_rad))
+        break
+    else:
+        print("\nEnter 1, 2 or 3")
+        continue
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW07_Task2.py:40]: nice

File: HW/Oleksandr Liashenko/HW07_Task3.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+word = input("Enter word: ")
+result = dict()
+for i in word:
+    dictionary = dict.fromkeys(i, word.count(i))
+    result.update(dictionary)
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW07_Task3.py:6]: in this case we dont need result dict

File: HW/Oleksandr Liashenko/HW08_Task1/main.py [hunk 13]
```diff
@@ -0,0 +1,5 @@
+from utils import *
+from models import *
+
+
+print(list(filter(lambda str: not ("__" in str), dir())))
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW08_Task1/main.py:5]: good

File: HW/Oleksandr Liashenko/HW08_Task2.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+digits = re.findall("[0-9]", password)
+lower = re.findall("[a-z]", password)
+upper = re.findall("[A-Z]", password)
+spec = re.findall("[@#$]", password)
+
+if digits and lower and upper and spec and 6 <= len(password) <= 16:
+  print("The password is valid")
+else:
+  print("The password is invalid")
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW08_Task2.py:13]: good

File: HW/Oleksandr Liashenko/HW08_task3/main.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    side_a = int(input("Enter side a length: "))
+    side_b = int(input("Enter side b length: "))
+    print("Area:", calculation.s_rectangle(side_a, side_b))
+elif figure == 2:
+    height = int(input("Enter height: "))
+    side_a = int(input("Enter side a length: "))
+    print("Area:", calculation.s_triangle(height, side_a))
+elif figure == 3:
+    radius = int(input("Enter radius: "))
+    print("Area:", round(calculation.s_circle(radius), 2))
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW08_task3/main.py:18]: good

File: HW/Oleksandr Liashenko/HW09_Task1.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+tries = 10
+
+while tries > 0:
+    usr_num = int(input(f"Enter a number ({tries} left): "))
+    if usr_num == rand_num:
+        print("Congratulations!")
+        break
+    elif tries == 1:
+        print("Game over")
+        break
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW09_Task1.py:13]: Dont understand this block, anyway while will be stoped

File: HW/Oleksandr Liashenko/HW09_Task2.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        COORD_Y = COORD_Y - DELTA_STEP
+    if keys[pygame.K_DOWN] and COORD_Y + DELTA_STEP + HEIGHT_RECTANGLE <= HEIGHT_DISPLAY:
+        COORD_Y = COORD_Y + DELTA_STEP
+
+    gameDisplay.fill(BLACK_COLOR)
+
+    pygame.draw.rect(gameDisplay, RED_COLOR, [COORD_X,
+                                              COORD_Y,
+                                              WIDTH_RECTANGLE,
+                                              HEIGHT_RECTANGLE])
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW09_Task2.py:49]: good

File: HW/Oleksandr Liashenko/HW09_Task3/OWM_module.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    observation = mgr.weather_at_place(city)
+    w = observation.weather
+
+    output_text = f"City: {city}\nConditions: {w.detailed_status}\nTemperature is " \
+                  f"{round(w.temperature('celsius')['temp'], 2)} C"\
+                  f"\nWind speed is {w.wind()['speed']} km/hour\nHumidity of the air is {w.humidity}"
+
+    return output_text
+
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW09_Task3/OWM_module.py:22]: nice

File: HW/Oleksandr Liashenko/HW09_Task3/Tk_OWM.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+# OWM_module.weather_response(entry_field.get())
+button.place(relx=0.7, rely=0, relwidth=0.3, relheight=1)
+
+lower_frame = tk.Frame(root, bg='gold', bd=10)
+lower_frame.place(relx=0.5, rely=0.25, relwidth=0.75, relheight=0.6, anchor='n')
+
+label = tk.Label(lower_frame, font=('Courier', 14))
+label.place(relx=0, rely=0, relwidth=1, relheight=1)
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW09_Task3/Tk_OWM.py:39]: good

File: HW/Oleksandr Liashenko/HW06_Task1.py (hunk 8 above)
Comment (oleksandr-liashenko) [reply HW/Oleksandr Liashenko/HW06_Task1.py:13]: Agree, my mistake. I should change "if i % 3 == 0" for "elif i % 3 == 0"

File: HW/Oleksandr Liashenko/HW07_Task3.py (hunk 12 above)
Comment (oleksandr-liashenko) [reply HW/Oleksandr Liashenko/HW07_Task3.py:6]: According to the task, the expected output is: {"h":1, "e":1, "l":2, "o":1}. Is there another way to get it?

Comment (ISierov) [reply HW/Oleksandr Liashenko/HW07_Task3.py:6]:
```
word = input("Enter word: ")
for i in word:
    dictionary = dict.fromkeys(i, word.count(i))
```
and here - dictionary already answer

File: HW/Oleksandr Liashenko/HW09_Task1.py (hunk 16 above)
Comment (oleksandr-liashenko) [reply HW/Oleksandr Liashenko/HW09_Task1.py:13]: This is a weird part, but it prevents printing "Try a greater number" or "Try a lesser number" in the case when this try is the last one (if I move this block outside the 'while' loop)

File: HW/Oleksandr Liashenko/HW10_Task1.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        print("This is a triangle")
+
+    def find_area(self):
+        s = (sum(self.sides)) / 2
+        print((s * (s-self.sides[0]) * (s-self.sides[1]) * (s-self.sides[2])) ** 0.5)
+
+
+triangle = Triangle()
+triangle.input_sides()
+triangle.find_area()
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW10_Task1.py:26]: Nice

File: HW/Oleksandr Liashenko/HW10_Task3.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+emp1.display()
+emp2.display()
+emp3.display()
+
+print(Employee.__base__.__subclasses__())
+print(Employee.__dict__)
+print(Employee.__name__)
+print(Employee.__module__)
+print(Employee.__doc__)
+
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW10_Task3.py:31]: Good

File: HW/Oleksandr Liashenko/HW10_Task2.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+
+    @staticmethod
+    def message():
+        print("Just a message")
+
+
+person = Human("Oleksandr")
+person.welcome()
+person.species()
+person.message()
```
Comment (ISierov) [inline HW/Oleksandr Liashenko/HW10_Task2.py:19]: Nice

**Pull request merged**